workflow-manager/task/task.go

```diff
@@ -0,0 +1,222 @@
+// in GCP PubSub. If dryRun is true, no tasks will actually be enqueued. Clients
+// should re-use a single instance as much as possible to enable batching of
+// publish requests.
+func NewGCPPubSubEnqueuer(ctx context.Context, project string, topicID string, dryRun bool) (*GCPPubSubEnqueuer, error) {
+	client, err := pubsub.NewClient(ctx, project)
+	if err != nil {
+		return nil, fmt.Errorf("pubsub.NewClient: %w", err)
+	}
+
+	return &GCPPubSubEnqueuer{
+		topic:   client.Topic(topicID),
+		context: ctx,
+		dryRun:  dryRun,
+	}, nil
+}
+
+func (e *GCPPubSubEnqueuer) Enqueue(task Task, completion func(error)) {
+	e.waitGroup.Add(1)
+	go func(ctx context.Context, task Task) {
+		defer e.waitGroup.Done()
+		jsonTask, err := task.MarshalToJson()
+		if err != nil {
+			completion(fmt.Errorf("marshaling task to JSON: %w", err))
+			return
+		}
+
+		if e.dryRun {
+			log.Printf("dry run, not enqueuing task")
+			completion(nil)
+			return
+		}
+
+		// Publish() returns immediately, giving us a handle to the result that we
+		// can block on to see if publishing succeeded. The PubSub client
+		// automatically retries for us, so we just keep the handle so the caller
+		// can do whatever they need to after successful publication and we can
+		// block in Stop() until all tasks have been enqueued
+		res := e.topic.Publish(e.context, &pubsub.Message{Data: jsonTask})
+		if _, err := res.Get(e.context); err != nil {
+			completion(fmt.Errorf("Failed to publish task %+v: %w", task, err))
+		}
+
+		completion(nil)
+	}(e.context, task)
+}
+
+func (e *GCPPubSubEnqueuer) Stop() {
+	e.waitGroup.Wait()
+}
+
+// AWSSNSEnqueuer implements Enqueuer using AWS SNS
+type AWSSNSEnqueuer struct {
+	service  *sns.SNS
+	topicARN string
+	dryRun   bool
+}
+
+func NewAWSSNSEnqueuer(region, identity, topicARN string, dryRun bool) (*AWSSNSEnqueuer, error) {
+	session, config, err := utils.AWSClientConfig(region, identity)
+	if err != nil {
+		return nil, err
+	}
+
+	return &AWSSNSEnqueuer{
+		service:  sns.New(session, config),
+		topicARN: topicARN,
+		dryRun:   dryRun,
+	}, nil
+}
+
+func (e *AWSSNSEnqueuer) Enqueue(task Task, completion func(error)) {
+	jsonTask, err := task.MarshalToJson()
+	if err != nil {
+		completion(fmt.Errorf("marshaling task to JSON: %w", err))
+		return
+	}
+
+	if e.dryRun {
+		log.Printf("dry run, not enqueuing task")
+		completion(nil)
+		return
+	}
+
+	// sns.Publish() blocks until the message has been saved by SNS, so no need
+	// to asynchronously handle completion.
+	// There's nothing in the PublishOutput we care about, so we discard it.
+	_, err = e.service.Publish(&sns.PublishInput{
+		TopicArn: aws.String(e.topicARN),
+		Message:  aws.String(string(jsonTask)),
+	})
+	if err != nil {
+		completion(fmt.Errorf("failed to publish task %+v: %w", task, err))
+		return
+	}
+
+	completion(nil)
+}
+
+func (e *AWSSNSEnqueuer) Stop() {
+	// No-op since sns.Publish() blocks until completion or failure
```

Comment:
In theory a future refactor could make it so that `Enqueue` is called concurrently in the main thread. In that situation, there could be inflight `Enqueue` calls when `.Stop()` is called, and this would be incorrect. I think it makes sense to go ahead and add a WaitGroup to `AWSNSEnqueue`, even though you'll be doing `.Add(1)` and `.Done()` in the same function with no goroutines.